chia/wallet/puzzles/load_clvm.py

```diff
@@ -1,53 +1,106 @@
 import importlib
 import inspect
 import os
+import time
+
+from typing import Optional, TextIO
 import pathlib
 
 import pkg_resources
-from clvm_tools.clvmc import compile_clvm as compile_clvm_py
 from chia.types.blockchain_format.program import Program, SerializedProgram
+from clvm_tools_rs import compile_clvm as compile_clvm_rust
+
+
+# Cribbed mostly from chia/daemon/server.py
+def create_exclusive_lock(lockfile) -> Optional[TextIO]:
+    """
+    Open a lockfile exclusively.
+    """
+
+    try:
+        fd = os.open(lockfile, os.O_CREAT | os.O_EXCL | os.O_RDWR)
+        f = open(fd, "w")
 
-compile_clvm = compile_clvm_py
+        f.write("lock")
+    except IOError:
+        return None
 
-# Handle optional use of clvm_tools_rs if available and requested
-if "CLVM_TOOLS_RS" in os.environ:
+    return f
+
+
+compile_clvm_py = None
+
+
+def translate_path(p_):
+    p = str(p_)
+    if os.path.isdir(p):
+        return p
+    else:
+        module_object = importlib.import_module(p)
+        return os.path.dirname(inspect.getfile(module_object))
+
+
+# Handle optional use of python clvm_tools if available and requested
+if "CLVM_TOOLS" in os.environ:
     try:
+        from clvm_tools.clvmc import compile_clvm as compile_clvm_py_candidate
 
+        compile_clvm_py = compile_clvm_py_candidate
+    finally:
+        pass
+
+
+def compile_clvm_in_lock(full_path, output, search_paths):
+    # Compile using rust (default)
+
+    # Ensure path translation is done in the idiomatic way currently
+    # expected.  It can use either a filesystem path or name a python
+    # module.
+    treated_include_paths = list(map(translate_path, search_paths))
+    res = compile_clvm_rust(str(full_path), str(output), treated_include_paths)
+
+    if "CLVM_TOOLS" in os.environ and os.environ["CLVM_TOOLS"] == "check" and compile_clvm_py is not None:
+        # Simple helper to read the compiled output
         def sha256file(f):
             import hashlib
 
             m = hashlib.sha256()
-            m.update(open(f).read().encode("utf8"))
+            m.update(open(f).read().strip().encode("utf8"))
             return m.hexdigest()
 
-        from clvm_tools_rs import compile_clvm as compile_clvm_rs
+        orig = "%s.orig" % output
+
+        compile_clvm_py(full_path, orig, search_paths=search_paths)
+        orig256 = sha256file(orig)
+        rs256 = sha256file(output)
+
+        if orig256 != rs256:
+            print("Compiled %s: %s vs %s\n" % (full_path, orig256, rs256))
+            print("Aborting compilation due to mismatch with rust")
+            assert orig256 == rs256
```

Comment:
I think it would be good to raise an exception explicitly here, rather than the assert. I believe asserts primarily are meant to document invariants. If this happens, we would want to exit, even if we're running in "release" mode.

(our codebase doesn't follow this rule, and sometimes use assert as input validation, but I think we should move away from that use of asserts)

Reply:
It's not meant precisely as input validation and is intended to be transient while people settle in (basically to be able to ask: is this really the same as it was).  Assert fit my mental model because it really does indicate that the world has gone bad; nothing else can be expected to work right.  I will consider how to do this better.

Reply:
yeah, that's a good point. This isn't really code that runs in production